### ISSUE TITLE
Investigate a criteria how to prioritize who is being shown on screen and who isn't

### ISSUE THREAD

**Issue by fpyoung** [montevideo-tech/hybrid-meetup-platform#118]
Since there is a limited space in terms of what cameras can be shown we need a way to prioritize who is being shown.
Those who are speaking should be always shown on screen. 
The main scope of this issue is to come up with a criteria to prioritize who deserves to be on the screen.

**Comment by mpollingerQualabs** [comment]
Issue #85 and PR #112 could provide some light on how to get who's talking and display those.

**Comment by marcovidonis** [comment]
@nreyes-qualabs had some ideas about this! It revolved around having an array of participants we subscribe to, ordered by priority. When we subscribe to a new participant, it will be inserted in the middle of the array, and the last item will be popped out.
Then the array can be reordered if we want to assign higher priority to some specific participant (for example, speakers, hosts and crowd camera should always be top priority).

**Comment by fpyoung** [comment]
okay so based on what Marco mentioned we suggest the approach to be:
- Create the array with priorities that will be called: displayedParticipants, this array will be filled with the first x participants that have the camera on initially.
- When someone speaks is going to be added in the array and have priority 1 and the array will be sorted.
- When someone turns on the camera, turns on the mic, joins the room, will be added to the array and have priority 2 and the array will be sorted.
Disclaimer: every time we add an element into the array we are going to remove the last element of the array.

**Comment by fpyoung** [comment]
Okay so as a first approach we are going to try to always show the people that are speaking on the screen, they are going to be the ones with higher priority https://github.com/montevideo-tech/hybrid-meetup-platform/issues/139 .
Then after this being solved we are going to show on screen with higher priority the ones that more recently have spoken, this is a 
requirement requested by @nicolaslevy . https://github.com/montevideo-tech/hybrid-meetup-platform/issues/140
Then after doing this add priority to the ones that have the camera on.
@marcovidonis what do you think about this approach? Keep in mind that the implementation is going to be the one that you mentioned on the first comment. I'm trying to split up the problem so the scope of the issues doesn't get that big.

**Comment by marcovidonis** [comment]
I agree with the approach, but let's not forget that we need to also change priorities based on the _role_ of each participant. For example, "crowd camera" will probably never have audio, but we want it to be visible whenever possible, because that will help give the feeling of a hybrid event. Remember that we're not making a team call, our project is much more than that :)

I think a way of handling this would be to assign a priority weight to each participant, and change this weight based on:
- are they currently speaking
- have they spoken recently
- their role

and then reorder the array based on priority weight. I imagine finding the perfect way of assigning weight will take some trials, but we can start simply by incorporating something like this (mine is just a suggestion!) and then work on it in time.

There's a [new blog from Daily](https://www.daily.co/blog/selecting-and-switching-participants-in-a-video-call-or-live-stream/) that talks about the challenges of defining which participants should be shown at any time in complex projects like ours. I think they really give some good points to think about.

@nicolaslevy @fyoungqualabs

**Comment by fpyoung** [comment]
Awesome we just read the article and it gave us some really interesting prompts to start working.
We are going to follow your guideline in terms of prioritization so heren is how it looks with issues and how are we going to develop them in the following order:
- https://github.com/montevideo-tech/hybrid-meetup-platform/issues/142 Implementing the priority queue.
- #119 since we don't have subtasks I implemented this issue and added in a comment which 'sub-issues' need to be done to have this issue completed.
Sub issues: of #119  
- https://github.com/montevideo-tech/hybrid-meetup-platform/issues/140
- https://github.com/montevideo-tech/hybrid-meetup-platform/issues/143

Next steps in prioritization:
- https://github.com/montevideo-tech/hybrid-meetup-platform/issues/144 This is an approximation to the solution since we haven't taken a look at what you guys did in terms of roles and what's left yet.

@marcovidonis let me know if you agree with this approach so I can close this issue.(my OCD is killing me 😆 )

**Comment by marcovidonis** [comment]
Sounds good @fyoungqualabs! Good plan! :)
